Adding Event Display macro

abragagn opened the pull request:
PR to add the event display macro EvDisp.

**Minimal instructions**
EvDisp works only in the DTSliceTestOfflineAnalysis framework.

To load it:
```
root -l
.L DTNtupleBaseAnalyzer.C++ 
.L EvDisp.C++ 
```

To launch it:
`auto evtDisplay = EvDisp(inputFile);`
Every time the constructor is called the following instructions are called:
```
INSTRUCTIONS

EvDisp::Loop(); // all events 
EvDisp::Loop(evt_number); // one event
EvDisp::LoopEntry(entry); // one entry
EvDisp::Loop(start, stop); // entries range
EvDisp::Loop(start, stop, evt); // one event, searched only in the entries range

EvDisp::DumpON(); // digi dump to screen ON 
EvDisp::DumpOFF(); // digi dump to screen OFF (default)

EvDisp::SaveDisplayON(); // save all .png (no promt) 
EvDisp::SaveDisplayPROMT(); // ask to save at every .png (default)
EvDisp::SaveDisplayOFF(); // do not save any .png (no promt) 

EvDisp::AskContinueON(); // ask if continue at every event ON (default) 
EvDisp::AskContinueOFF(); // ask if continue at every event OFF

EvDisp::RunOnlyIfOneEmptyON(); // run only on events where one of the two Phases has no hits 
EvDisp::RunOnlyIfOneEmptyOFF(); // disable RunOnlyIfOneEmpty (default)

```

**Functions**
EvDisp create a canvas with the schematic representations of the four MB station for the legacy readout (left) and Phase 2 (right). Digi hits are shown as well as segments.
Multiple hits in the same wires are coded with the following color scheme:
1. single hits: kGreen
2. double hits: kYellow
3. triple hits: kOrange
4. quadruple hits: kRed
5. quintuple (or higher) hits: kBlue

**User control**
The user can:
- run on a single entry
- run on a single evt number
- run on a entry range
- run over all entries

The user can decide on per--event basis if
- save the display as a .png
- to go to the next event

these promts can be bypassed (see instructions).
Digi information dump is available.
For more information see instructions.

**Hardcoded stuffs**
- Display runs only on wheel 2, sector 12
- The code run on all stations but flags to skip each station are available and initialized in the constructor: _vetoMB1_, _vetoMB2_, _vetoMB3_, _vetoMB4_.. Currently **only** _vetoMB2_ is false (i.e. **only** station 2 is displayed). No function are provided to change these variables
- Code to draw phase 2 Segments is commented since no ph2Segments are available.


**Known issue**
- Object created with DrawClone are not released from memory.
- Same layer staggering for all 4 station.  Suggestion: to implement any difference modify the function computeX(float wire, int layer, int iMB)

battibass commented on analyses/EvDisp.C:
```diff
@@ -0,0 +1,628 @@
+{
+
+  // STRUCT DRAWING
+  auto c1 = new TCanvas("c1","c1",1500,1000);
```
Please reduce the 1500 - 1000 canvas size to 1200 - 800.
A bigger canvas causes problems, at least on MACs.

Also, if the canvas could be created at the beginning (e.g. in book) and re-filled every event, instead of being deleted and recreated, this would make the focus of the prompt remain in the root prompt and make use easier.
This last feature could be implemented in a future version though.

battibass replied:
Both requests were implemented and things appear to be working fine

battibass commented:
Besides the specific comments on the canvas, I tried to run:
```
auto evtDisplay = EvDisp("/eos/cms/store/group/dpg_dt/comm_dt/commissioning_2019_data/ntuples/DTDPGNtuple_run330463.root")
evtDisplay.Loop(1698)
```
and I get a crash.
Could you have a look at this before I integrate the PR?

abragagn commented:
Comments on last commits:

- Moved canvas creation and graphStruct drawing to Book(), now graphStruct is not redrawn at every events, execution should be faster
- Removed all the DrawClones() calls, instead opted for adding additional dimensions (i.e. station index) to the objects vectorization. This allows to a better memory management and overall fast drawing.

battibass commented:
The issue reported [earlier in this thread](https://github.com/battibass/DTSliceTestOfflineAnalysis/pull/2#pullrequestreview-271416426) has been fixed.
I will now merge the PR into the master branch, so that more people can start experimenting with the event display.
It would be great if a list of the known updates to be performed (e.g. SL staggering for all 4 MB stations etc ...) could be added as issues to this repository.
